addons/mass_mailing/models/mass_mailing.py

```diff
@@ -644,19 +721,28 @@ def _get_opt_out_list(self):
         self.ensure_one()
         opt_out = {}
         target = self.env[self.mailing_model_real]
-        mail_field = 'email' if 'email' in target._fields else 'email_from'
-        if 'opt_out' in target._fields:
-            # avoid loading a large number of records in memory
-            # + use a basic heuristic for extracting emails
-            query = """
-                SELECT lower(substring(%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)'))
-                  FROM %(target)s
-                 WHERE opt_out AND
-                       substring(%(mail_field)s, '([^ ,;<@]+@[^> ,;]+)') IS NOT NULL;
-            """
-            query = query % {'target': target._table, 'mail_field': mail_field}
-            self._cr.execute(query)
-            opt_out = set(m[0] for m in self._cr.fetchall())
+        [mail_field] = target.primary_email
+        if 'opt_out' in target._fields or 'opt_out_list_ids' in target._fields:
```

Comment:
To squash with other commit; I would also probably make an SQL query to avoid having a browse + mapped on contact_id which is probably costly.